### ISSUE TITLE
`*caller` changes depending some unknown reason with jRuby 9.3

### ISSUE THREAD

**Issue by aurelien-reeves** [jruby/jruby#6858]
**Environment Information**
  
Provide at least:
* JRuby version: `jruby 9.3.0.0 (2.6.8) 2021-09-17 85c20e780f OpenJDK 64-Bit Server VM 11.0.11+9-post-Debian-1deb10u1 on 11.0.11+9-post-Debian-1deb10u1 +jit [linux-x86_64]`
* Operating system and platform: `Linux ***** 5.10.16.3-microsoft-standard-WSL2 #1 SMP Fri Apr 2 22:23:49 UTC 2021 x86_64 GNU/Linux`

Refs. https://github.com/cucumber/cucumber-ruby/pull/1571

With jRuby 9.3, test suite of cucumber-ruby is failing. There is a specific case that cannot be reduced to a single test where an extra entry appears in the caller stack.

To reproduce, you can checkout cucumber-ruby branch [`fix-jruby-9.3`](https://github.com/cucumber/cucumber-ruby/tree/fix-jruby-9.3) then execute

    bundle exec rspec './spec/cucumber/formatter/junit_spec.rb[1:4:3:5:1,1:4:3:5:2,1:4:3:6:1:1,1:4:3:6:2:1,1:4:3:6:3:1,1:4:3:6:4:1,1:5:1:1:1,1:5:1:1:2,1:5:1:1:3,1:5:1:1:4,1:5:1:1:5,1:5:1:1:6,1:5:1:1:7]' './spec/cucumber/formatter/pretty_spec.rb[1:1:1:9:1,1:1:1:10:1,1:5:3:1,1:5:4:1]' './spec/cucumber/formatter/progress_spec.rb[1:1:4:1:1,1:1:4:2:1,1:1:4:3:1]' './spec/cucumber/formatter/query/step_definitions_by_test_step_spec.rb[1:1:1:1:1,1:1:1:2:2:1,1:1:2:1:1,1:1:2:2:1]' './spec/cucumber/glue/proto_world_spec.rb[1:2:1:1,1:2:2:1,1:2:3:1,1:2:4:1,1:3:1:1,1:3:1:2]' './spec/cucumber/glue/step_definition_spec.rb[1:1,1:2,1:3:1,1:3:2,1:3:3,1:3:4,1:4,1:5,1:6,1:7,1:8,1:9,1:10,1:13,1:14]'

You will be able to note that there is an extra line only with jruby 9.3 in the logs: `lib/cucumber/glue/step_definition.rb:26`. As soon as you do not execute that exact test-suite, or use another ruby implementation, that extra caller disappear and the tests are passing

**Expected Behavior**

tests are passing

**Actual Behavior**

1 test is failing

![image](https://user-images.githubusercontent.com/24386780/134888473-9cb9cf98-276d-4ef0-9b18-d4b3984c1c62.png)

**Comment by headius** [comment]
This is a dupe of #6846, which I will fix today!

**Comment by headius** [comment]
I tried to confirm that your issue is fixed by #6877 but got the following error bundling:

```
$ bundle install
Fetching https://github.com/cucumber/cucumber-ruby-core.git
Fetching https://github.com/cucumber/cucumber-ruby-wire.git
Ignoring jruby-launcher-1.1.18-java because its extensions are not built. Try: gem pristine jruby-launcher --version 1.1.18
Fetching gem metadata from https://rubygems.org/.........
Resolving dependencies......
Bundler could not find compatible versions for gem "cucumber-gherkin":
  In Gemfile:
    cucumber was resolved to 7.0.0, which depends on
      cucumber-gherkin (~> 20.0, >= 20.0.1)

    cucumber-core was resolved to 10.0.1, which depends on
      cucumber-gherkin (~> 22.0, >= 22.0.0)
```

The fix has landed on JRuby master (9.3.1.0) and you could try it with a snapshot from https://oss.sonatype.org/content/repositories/snapshots/org/jruby/jruby-dist/9.3.1.0-SNAPSHOT/ (once master builds catch up with the PR).

**Comment by aurelien-reeves** [comment]
Thanks a lot!

I have updated the branch on cucumber-ruby to fix the `bundle install` issue.

I also have executed the tests with the last jruby snapshot available in the link you gave, it is working fine ! \o/

Thanks a lot 👌
